src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java

```diff
@@ -124,11 +126,23 @@ static <T> T executeWithRetries(
       Function<T, Long> statusGetter)
       throws IOException, IngestResponseException {
     int retries = 0;
-    T response;
+    int authRetries = 0;
+    T response = null;
     do {
       try (CloseableHttpResponse httpResponse =
           httpClient.execute(
               requestBuilder.generateStreamingIngestPostRequest(payload, endpoint, message))) {
+
+        // SNOW-877039 in case token is expired
+        if (httpResponse.getStatusLine().getStatusCode() == HttpStatus.SC_UNAUTHORIZED) {
+          if (authRetries < MAX_TOKEN_REFRESH_FOR_UNAUTHORIZE_RETRY) {
+            requestBuilder.refreshToken();
```

Comment:
Lets add a logger info here

Reply:
Can we implement a retry in `handleExceptionalStatus`? that will benefit other apis as well (snowpipe)

Reply:
Sounds good, moved into `handleExceptionalStatus`